// Copyright (c) 2022 Robert Bosch GmbH and Microsoft Corporation
//
// This program and the accompanying materials are made available under the
// terms of the Apache License, Version 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations
// under the License.
//
// SPDX-License-Identifier: Apache-2.0


syntax = "proto3";

package sdv.edge.comfort.hvac.v1;

/**
 * @brief Example HVAC service for controlling the heating, ventilation, and air
          conditioning elements of the vehicle cabin.
 *        This definition is designed here according to the draft of the comfort seats
 *        service definition of the COVESA Vehicle Service Catalog (VSC) (see
 *        https://github.com/COVESA/vehicle_service_catalog) as a definition of an
 *        HVAC service is currently missing in VSC.
 */
service Hvac {
    /** Set the desired ac status
     *
     *  Returns gRPC status codes:
     *   * OK - AcStatus set
     *   * INVALID_ARGUMENT - The requested AcStatus is not supported by the service instance
     *   * INTERNAL - A HVAC service internal error happened - see error message for details
    */
    rpc SetAcStatus(SetAcStatusRequest) returns (SetAcStatusReply);

    /** Set the desired cabin temparature
     *
     *  Returns gRPC status codes:
     *   * OK - Desired temperature set
     *   * OUT_OF_RANGE - The specified temperature is not supported in this vehicle
     *   * INTERNAL - A HVAC service internal error happened - see error message for details
    */
    rpc SetTemperature(SetTemperatureRequest) returns (SetTemperatureReply);

}

/**
 * @brief
 *
 */
message SetAcStatusRequest {
    AcStatus status = 1; // The desired status of A/C
}

/**
 * @brief
 *
 */
message SetAcStatusReply {}

/**
 * @brief
 *
 */
message SetTemperatureRequest {
    float temperature = 1; // The desired cabin temperature in degree Celsius
}

/**
 * @brief
 *
 */
message SetTemperatureReply {}

enum AcStatus {
    OFF = 0;
    ON  = 1;
}
